classes/external/trigger_dataflow.php

```
@@ -0,0 +1,77 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_dataflows\external;
+
+use external_function_parameters;
+use external_single_structure;
+use external_value;
+use tool_dataflows\dataflow;
+use tool_dataflows\task\process_dataflow_ad_hoc;
+
+/**
+ * Trigger dataflow webservice function.
+ *
+ * @package    tool_dataflows
+ * @author     Kevin Pham <kevinpham@catalyst-au.net>
+ * @copyright  Catalyst IT, 2023
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class trigger_dataflow extends \external_api {
+
+    /**
+     * Returns description of method parameters
+     * @return external_function_parameters
+     */
+    public static function execute_parameters() {
+        return new external_function_parameters([
+            'dataflow' => new external_single_structure([
+                'id' => new external_value(PARAM_INT, 'dataflow record id'),
+            ])
+        ]);
+    }
+
+    /**
+     * Returns description of method return values
+     * @return external_single_structure
+     */
+    public static function execute_returns() {
+        return new external_single_structure([
+            'id' => new external_value(PARAM_INT, 'dataflow record id'),
+        ]);
+    }
+
+    /**
+     * Trigger a dataflow run
+     *
+     * Currently it only supports asynchronous execution.
+     *
+     * @param array $dataflow array of dataflow run information.
+     * @return array of newly created groups
+     */
+    public static function execute($dataflow) {
+        $params = self::validate_parameters(self::execute_parameters(), ['dataflow' => $dataflow]);
+
+        $dataflow = new dataflow($params['dataflow']['id']);
```

my gut feel here is that while we dont need a trigger specifically for it, I still want to have one I think. I like the idea that you can't execute an arbitrary dataflow, ONLY ones that opt into it via a webservice trigger. So basically, if step 0 isnt a webservice trigger, just throw a warning and do nothing.